src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticAnalyzerAPIUsageAnalyzer.cs

```diff
@@ -0,0 +1,200 @@
+
+                        if (violatingTypeNamesBuilder.Count > 0)
+                        {
+                            string[] args;
+                            DiagnosticDescriptor rule;
+                            if (violatingUsedTypeNamesBuilder.Count == 1 && violatingUsedTypeNamesBuilder.Single() == declaredType.ToDisplayString())
+                            {
+                                // Change diagnostic analyzer type '{0}' to remove all direct accesses to type(s) '{1}'
+                                rule = DoNotUseTypesFromAssemblyDirectRule;
+                                args = new[]
+                                {
+                                    declaredType.ToDisplayString(SymbolDisplayFormat.MinimallyQualifiedFormat),
+                                    string.Join(", ", violatingTypeNamesBuilder)
+                                };
+                            }
+                            else
+                            {
+                                // Change diagnostic analyzer type '{0}' to remove all direct and/or indirect accesses to type(s) '{1}', which access type(s) '{2}'
+                                rule = DoNotUseTypesFromAssemblyIndirectRule;
+                                args = new[]
+                                {
+                                    declaredType.ToDisplayString(SymbolDisplayFormat.MinimallyQualifiedFormat),
+                                    string.Join(", ", violatingUsedTypeNamesBuilder),
+                                    string.Join(", ", violatingTypeNamesBuilder)
+                                };
+                            }
+
+                            Diagnostic diagnostic = declaredType.CreateDiagnostic(rule, args);
+                            compilationEndContext.ReportDiagnostic(diagnostic);
+                        }
+                    }
+                });
+            });
+        }
+
+        private ImmutableHashSet<INamedTypeSymbol> GetUsedNamedTypes(INamedTypeSymbol namedType, Compilation compilation, CancellationToken cancellationToken, ref bool hasAccessToTypeFromWorkspaceAssemblies)
+        {
+            var builder = ImmutableHashSet.CreateBuilder<INamedTypeSymbol>();
+            foreach (var decl in namedType.DeclaringSyntaxReferences)
+            {
+                var syntax = decl.GetSyntax(cancellationToken);
+
+                // GetSyntax for VB returns the StatementSyntax instead of BlockSyntax node.
+                syntax = syntax.FirstAncestorOrSelf<SyntaxNode>(node => IsNamedTypeDeclarationBlock(node), ascendOutOfTrivia: false);
+
+                Func<SyntaxNode, bool> descendantIntoChildren = node => ReferenceEquals(node, syntax) || !IsNamedTypeDeclarationBlock(node);
+                var semanticModel = compilation.GetSemanticModel(syntax.SyntaxTree);
+                foreach (TTypeSyntax typeSyntax in syntax.DescendantNodes(descendantIntoChildren).OfType<TTypeSyntax>())
```

Comment:
can we use IOperation instead of using Syntax?

Reply:
shouldn't this check all expression not just type syntax?

Reply:
so, since it is checking all executable code, can we get Ioperation tree for each code block, (of diagnostic analzyer) and visit all Ioperation and see its type is from those workspace assemblies?

Reply:
@heejaechang Thank you! I figured out that we need to visit operation tree for implicit type references, which was not covered by my previous implementation. However, for executable code too there are bunch of type syntax nodes for which we have no operations. For example, VB method block returns the operation tree, but there is no operation for type syntax nodes for parameters or return type. Lambda and local function declarations do not have specific operation nodes for parameters or return type either.

So for simplicity, I have taken the following approach:
1. Visit all operations in executable code and handle `IOperation.Type` for these nodes - no need to handle individual `OperationKind` (we need to handle static member accesses specially as they have no instance operation).
2. Visit all explicit `TypeSyntax` descendants in the named type declaration - covers types explicitly mentioned in declarations and executable code.